<?xml version="1.0"?>
<robot xmlns:xacro="http://www.ros.org/wiki/xacro">

  <xacro:macro name="diffbot_ros2_control" params="name prefix use_mock_hardware">

    <ros2_control name="${name}" type="system">
      <xacro:unless value="${use_mock_hardware}">
        <hardware>
          <plugin>diffdrive_roboclaw/DiffDriveRoboclawHardware</plugin>
          <param name="left_front_wheel_joint">left_front_wheel_joint</param>
          <param name="left_back_wheel_joint">left_back_wheel_joint</param>
          <param name="right_front_wheel_joint">right_front_wheel_joint</param>
          <param name="right_back_wheel_joint">right_back_wheel_joint</param>
          <param name="loop_rate">30</param>
          <param name="device">/dev/ttyACM0</param>
          <param name="baud_rate">38400</param>
          <param name="timeout_ms">1000</param>
          <param name="enc_counts_per_rev">4096</param>
        </hardware>
      </xacro:unless>
      <xacro:if value="${use_mock_hardware}">
        <hardware>
          <plugin>mock_components/GenericSystem</plugin>
          <param name="calculate_dynamics">true</param>
        </hardware>
      </xacro:if>
      <joint name="${prefix}left_front_wheel_joint">
        <command_interface name="velocity">
          <param name="min">-0.5</param>
          <param name="max">0.5</param>
        </command_interface>
        <state_interface name="position" />
        <state_interface name="velocity" />
      </joint>
      <joint name="${prefix}right_front_wheel_joint">
        <command_interface name="velocity">
          <param name="min">-0.5</param>
          <param name="max">0.5</param>
        </command_interface>
        <state_interface name="position" />
        <state_interface name="velocity" />
      </joint>
      <joint name="${prefix}left_back_wheel_joint">
        <command_interface name="velocity">
          <param name="min">-0.5</param>
          <param name="max">0.5</param>
        </command_interface>
        <state_interface name="position" />
        <state_interface name="velocity" />
      </joint>
      <joint name="${prefix}right_back_wheel_joint">
        <command_interface name="velocity">
          <param name="min">-0.5</param>
          <param name="max">0.5</param>
        </command_interface>
        <state_interface name="position" />
        <state_interface name="velocity" />
      </joint>
    </ros2_control>

  </xacro:macro>

</robot>